### PR TITLE
#576: Close all the parts on request's body close

### PR DESCRIPTION
**Pull request by essobedo**
Fix for https://github.com/yegor256/takes/issues/576

The goal of this PR is to close all the parts of the request in order to ensure that all the temporary files are properly deleted.

### CODE REVIEW

**Comment by davvd** [comment]
@essobedo I will find a reviewer for your pull requests shortly, thanks for contribution!

**Comment by davvd** [comment]
@mkordas review this one, please

**Comment by mkordas** [comment]
@essobedo I'll do review today

**Comment by mkordas** [comment]
@essobedo I'm not sure if @davvd will understand [PR](#576) link, it's better to post just `#576`

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 1]
```diff
@@ -271,6 +272,100 @@ public void parsesHttpBody() throws IOException {
     }
 
     /**
+     * RqMultipart.Base can close all parts once the request body has been
+     * closed.
+     * @throws IOException If some problem inside
+     */
+    @Test(expected = IOException.class)
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo even if first line of the test below throws `IOException`, the test would still pass. We need more precise assertion here.

**Comment by essobedo** [comment]
@mkordas good to know but actually the link is more for the QA not really for @davvd

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 2]
```diff
@@ -271,6 +272,100 @@ public void parsesHttpBody() throws IOException {
...
+        MatcherAssert.assertThat(
+            multi.part("foo").iterator().next(),
+            Matchers.notNullValue()
+        );
+        multi.part("foo").iterator().next().body().close();
+        MatcherAssert.assertThat(
+            multi.part("bar").iterator().next(),
+            Matchers.notNullValue()
+        );
+        multi.part("bar").iterator().next().body().close();
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo don't we need any assertion after this line?

**Comment by mkordas** [comment]
@essobedo actually it is mainly for PM to check if there is any code created for the issue

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 3]
```diff
@@ -97,8 +98,8 @@
      * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
      * @see RqGreedy
      */
-    @EqualsAndHashCode(callSuper = true)
-    final class Base extends RqWrap implements RqMultipart {
+    @EqualsAndHashCode(of = "origin")
```
Comment (mkordas) [inline src/main/java/org/takes/rq/RqMultipart.java]: @essobedo why do we need to change this line?

Comment (mkordas) [inline src/main/java/org/takes/rq/RqMultipart.java]: @essobedo sorry, now I see - we removed inheritance

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 4]
```diff
@@ -351,6 +363,34 @@ private void copy(final WritableByteChannel target,
...
             return map;
         }
+
+        /**
+         * Decorator allowing to close all the parts of the request.
+         */
+        private class CloseMultipart extends FilterInputStream {
+
+            /**
+             * Creates a {@code CloseParts} with the specified {@code input}.
```
Comment (mkordas) [inline src/main/java/org/takes/rq/RqMultipart.java]: @essobedo I don't think we need to use `@code` in this sentence

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 5]
```diff
@@ -271,6 +272,100 @@ public void parsesHttpBody() throws IOException {
...
+                multi.part("content").iterator().next(),
+                Matchers.notNullValue()
+            );
+            multi.part("content").iterator().next().body().read();
+        }
+    }
+
+    /**
+     * RqMultipart.Base can close all parts explicitly even if the request body
+     * has been closed.
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo where is the assertion about parts being closed?

**Comment by mkordas** [comment]
@essobedo see my comments above

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 1 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas ok

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 2 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas Here the goal is to prove (for backward compatibility reason) that we don't get IOException when we close explicitly a part even after calling `multi.body().close()`. So not getting exception is somehow the assertion here.

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 3 above)
Comment (essobedo) [inline src/main/java/org/takes/rq/RqMultipart.java]: @mkordas exactly

File: src/main/java/org/takes/rq/RqMultipart.java (hunk 4 above)
Comment (essobedo) [inline src/main/java/org/takes/rq/RqMultipart.java]: @mkordas ok

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 5 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas I don't have access to the underlying Stream so the only way I have to ensure that it is closed is to try to read the stream and expect an IOException

**Comment by essobedo** [comment]
@mkordas thanks for your comments, please check my answers and the new version of the PR.

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 5 above)
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo OK, so we should make assetion on `IOException` message

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 2 above)
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo OK, can you put that information to Javadoc of the test?

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 1 above)
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo we need to remove `(expected = IOException.class)` from here

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 2 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas ok

File: src/test/java/org/takes/rq/RqMultipartTest.java (hunk 5 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas ok

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 6]
```diff
@@ -271,6 +272,109 @@ public void parsesHttpBody() throws IOException {
...
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void closesExplicitlyAllParts() throws IOException {
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo tests should declare `throws Exception`, not `throws IOException` - thats our coding style.

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 7]
```diff
@@ -271,6 +272,109 @@ public void parsesHttpBody() throws IOException {
     }
 
     /**
+     * RqMultipart.Base can close all parts once the request body has been
+     * closed.
+     * @throws IOException If some problem inside
+     */
+    @Test(expected = IOException.class)
+    public void closesAllParts() throws IOException {
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo same here, use `throws Exception`

**Comment by mkordas** [comment]
@essobedo see my responses

**Comment by essobedo** [comment]
@mkordas here is a new version of the PR, could you please check?

File: src/test/java/org/takes/rq/RqMultipartTest.java [hunk 8]
```diff
@@ -271,6 +272,126 @@ public void parsesHttpBody() throws IOException {
...
+                            "form-data; name=\"content\"; filename=\"a.bin\""
+                        )
+                    )
+                )
+            );
+            multi = new RqMultipart.Base(request);
+            multi.part("name").iterator().next().body().read();
+            multi.part("content").iterator().next().body().read();
+            multi.body().close();
+        } catch (final IOException ex) {
```
Comment (mkordas) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @essobedo do we need this catch now?

Comment (essobedo) [inline src/test/java/org/takes/rq/RqMultipartTest.java]: @mkordas good catch, it is indeed useless now.

**Comment by essobedo** [comment]
@mkordas PR updated, please recheck

**Comment by mkordas** [comment]
@essobedo looks all right!

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thank you :)

**Comment by davvd** [comment]
@mkordas 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 80018307... done, I added **34 mins** in payment `56e32ace3239de01a900032e`, 64 hours and 39 mins spent to complete this... review comments (c=19) added as a bonus... +34 added to your rating, current score is: [+8605](http://www.netbout.com/b/36080?open=rating)

**Pull request merged**